listenerutil/parse.go

```diff
@@ -387,3 +413,142 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
 		return "", fmt.Errorf("multiple addresses found (%q), please configure one", out)
 	}
 }
+
+const strictTransportSecurity = "max-age=31536000; includeSubDomains"
+const xContentTypeOptions = "nosniff"
+const cacheControl = "no-store"
+const apiContentSecurityPolicy = "default-src 'none'"
+const uiContentSecurityPolicy = "default-src 'none'; script-src 'self'; frame-src 'self'; font-src 'self'; connect-src 'self'; img-src 'self' data:*; style-src 'self'; media-src 'self'; manifest-src 'self'; style-src-attr 'self'; frame-ancestors 'self'"
+
+// parseCustomResponseHeaders takes raw config values for the "custom_ui_response_headers"
+// and "custom_api_response_headers". It makes sure the config entry is passed in as a map
+// of status code to a map of header name and header values. It verifies the validity of the
+// status codes, and header values. It also adds the default headers values for "Cache-Control",
+// "Strict-Transport-Security", "X-Content-Type-Options", and "Content-Security-Policy".
+func parseCustomResponseHeaders(responseHeaders interface{}, uiHeaders bool) (map[int]map[string][]string, error) {
+	h := make(map[int]map[string][]string)
+	// if responseHeaders is nil, we still should set the default custom headers
+	if responseHeaders == nil {
+		h[0] = map[string][]string{
+			"Strict-Transport-Security": {strictTransportSecurity},
+			"X-Content-Type-Options":    {xContentTypeOptions},
+			"Cache-Control":             {cacheControl},
+		}
+		if uiHeaders {
+			h[0]["Content-Security-Policy"] = []string{uiContentSecurityPolicy}
+		} else {
+			h[0]["Content-Security-Policy"] = []string{apiContentSecurityPolicy}
+		}
+		return h, nil
+	}
+
+	customResponseHeader, ok := responseHeaders.([]map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("response headers were not configured correctly. Please make sure they're in a list of maps")
+	}
+
+	for _, crh := range customResponseHeader {
+		for statusCode, responseHeader := range crh {
+			headerValList, ok := responseHeader.([]map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("response headers were not configured correctly. Please make sure they're in a list of maps")
+			}
+
+			status, err := convertStatusCode(statusCode)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(headerValList) != 1 {
+				return nil, fmt.Errorf("invalid number of response headers exist")
+			}
+			headerValMap := headerValList[0]
+			headerVal, err := parseHeaders(headerValMap)
+			if err != nil {
+				return nil, err
+			}
+
+			h[status] = headerVal
+		}
+	}
+
+	// setting default headers
+	if h[0] == nil {
```

Comment:
Up above if no values are set then you are setting defaults. At this point here you know that values have been set, but you are adding headers in if they were not set by the user anyway. This would make it impossible for users to _unset_ values we are setting by default that they don't want. I think this entire block probably should be removed.

Reply:
so I've added what I think is a happy medium. the defaults are still set if a user doesn't specify that specific header, but if the user specifies an empty header, it'll clear the defaults. this will also be reflected in the config test

Reply:
That makes sense to me